### ISSUE TITLE
Unable to install app

### ISSUE THREAD

**Issue by ValdasK** [nextcloud/mail#2383]
### Expected behavior
App should install without errors

### Actual behavior
When installing app, first it failed due invalid query. Something about having multiple primary keys. Now on install it says "The table with name 'nextcloud.oc_mail_mailboxes' already exists."

### Mail app

**Mail app version:** (see apps admin page, e.g. 0.5.3)
0.21.0

**Mailserver or service:** (e.g. Outlook, Yahoo, Gmail, Exchange,...)
not yet configured

### Server configuration

**Operating system**: (e.g. Debian 8)
17.0.1-fpm docker image

**Web server:** (e.g. Apache, Nginx,...)
nginx 

**Database:** (e.g. MariaDB, SQLite or PostgreSQL)
mariadb

**PHP version:** (e.g. 7.0)
7?

**Nextcloud Version:** (see admin page, e.g. 13.0.2)
17.0.1

#### Client configuration
**Browser:** (e.g. Firefox 48)
FF 71

**Operating system:** (e.g. Arch Linux)
any

**Comment by ChristophWurst** [comment]
Sounds like you had an older version of the app installed previously and/or you changed the DB schema manually. Is that the case?

**Comment by ValdasK** [comment]
I don't recall having Mail installed, unless it was by default. I have same instance since NC 9. Haven't logged to database manually ever. Do you think that removing table manually should help to solve this issue?

**Comment by ChristophWurst** [comment]
Okay. Could you check the `oc_migrations` table to see which of the migrations of this app were run? Hope that helps diagnose why you got the errors.

**Comment by ValdasK** [comment]
```
MariaDB [nextcloud]> select * from oc_migrations;
+-----------------------+--------------------------+
| app                   | version                  |
+-----------------------+--------------------------+
| activity              | 2006Date20170808154933   |
| activity              | 2006Date20170808155040   |
| activity              | 2006Date20170919095939   |
| activity              | 2007Date20181107114613   |
| activity              | 2008Date20181011095117   |
| activity              | 2010Date20190416112817   |
| core                  | 13000Date20170705121758  |
| core                  | 13000Date20170718121200  |
| core                  | 13000Date20170814074715  |
| core                  | 13000Date20170919121250  |
| core                  | 13000Date20170926101637  |
| core                  | 14000Date20180129121024  |
| core                  | 14000Date20180404140050  |
| core                  | 14000Date20180516101403  |
| core                  | 14000Date20180518120534  |
| core                  | 14000Date20180522074438  |
| core                  | 14000Date20180626223656  |
| core                  | 14000Date20180710092004  |
| core                  | 14000Date20180712153140  |
| core                  | 15000Date20180926101451  |
| core                  | 15000Date20181015062942  |
| core                  | 15000Date20181029084625  |
| core                  | 16000Date20190207141427  |
| core                  | 16000Date20190212081545  |
| core                  | 16000Date20190427105638  |
| core                  | 16000Date20190428150708  |
| core                  | 17000Date20190514105811  |
| dav                   | 1004Date20170825134824   |
| dav                   | 1004Date20170919104507   |
| dav                   | 1004Date20170924124212   |
| dav                   | 1004Date20170926103422   |
| dav                   | 1005Date20180413093149   |
| dav                   | 1005Date20180530124431   |
| dav                   | 1006Date20180619154313   |
| dav                   | 1006Date20180628111625   |
| dav                   | 1008Date20181030113700   |
| dav                   | 1008Date20181105104826   |
| dav                   | 1008Date20181105104833   |
| dav                   | 1008Date20181105110300   |
| dav                   | 1008Date20181105112049   |
| dav                   | 1008Date20181114084440   |
| dav                   | 1011Date20190725113607   |
| dav                   | 1011Date20190806104428   |
| dav                   | 1012Date20190808122342   |
| mail                  | 0100Date20180825194217   |
| mail                  | 0110Date20180825195812   |
| mail                  | 0110Date20180825201241   |
| mail                  | 0130Date20190408134101   |
| mail                  | 0156Date20190828140357   |
| mail                  | 0161Date20190902103559   |
| notifications         | 2004Date20190107135757   |
| oauth2                | 010401Date20181207190718 |
| oauth2                | 010402Date20190107124745 |
| previewgenerator      | 020000Date20180823071939 |
| privacy               | 100Date20190217131943    |
| spreed                | 2000Date20170707093535   |
| spreed                | 2000Date20171026140256   |
| spreed                | 2000Date20171026140257   |
| spreed                | 2001Date20170707115443   |
| spreed                | 2001Date20170913104501   |
| spreed                | 2001Date20170921145301   |
| spreed                | 2001Date20170929092606   |
| spreed                | 2001Date20171009132424   |
| spreed                | 2001Date20171026134605   |
| spreed                | 2001Date20171026141336   |
| spreed                | 2001Date20171031102049   |
| spreed                | 2001Date20180103144447   |
| spreed                | 2001Date20180103150836   |
| spreed                | 3002Date20180319104030   |
| spreed                | 3003Date20180707222130   |
| spreed                | 3003Date20180718112436   |
| spreed                | 3003Date20180718133519   |
| spreed                | 3003Date20180720162342   |
| spreed                | 3003Date20180722152733   |
| spreed                | 3003Date20180722152849   |
| spreed                | 3003Date20180730080327   |
| spreed                | 4099Date20181001123058   |
| spreed                | 5099Date20190121102337   |
| spreed                | 5099Date20190319134820   |
| spreed                | 6099Date20190627172429   |
| spreed                | 7000Date20190717141457   |
| spreed                | 7000Date20190724121136   |
| spreed                | 7000Date20190724121137   |
| text                  | 010000Date20190617184535 |
| twofactor_backupcodes | 1002Date20170607104347   |
| twofactor_backupcodes | 1002Date20170607113030   |
| twofactor_backupcodes | 1002Date20170919123342   |
| twofactor_backupcodes | 1002Date20170926101419   |
| twofactor_backupcodes | 1002Date20180821043638   |
+-----------------------+--------------------------+
```

**Comment by ChristophWurst** [comment]
Okay, so as far as I understand you didn't have the app installed but one of the migrations failed and thus this unexpected db state vs migrations is block from continuing.

Instead of just resetting everything and hoping for the best on a second try I would like to understand the problem better as it might also occur for other admins.

> mariadb

Could you tell me the exact versions? Then I'll fire a CI build with the same database to see if that works.

**Comment by ValdasK** [comment]
```
MariaDB [(none)]> select version();
+--------------------------+
| version()                |
+--------------------------+
| 10.1.38-MariaDB-0+deb9u1 |
+--------------------------+
```

What is interesting that if I check tables, I can see 
```
| oc_mail_accounts                |
| oc_mail_aliases                 |
| oc_mail_attachments             |
| oc_mail_coll_addresses          |
| oc_mail_mailboxes               |
```

even though migration `0161Date20190902103559` is marked as completed, which does `$schema->dropTable('mail_mailboxes');`

**Comment by ChristophWurst** [comment]
Bad news: 10.1 works fine: https://github.com/nextcloud/mail/pull/2401

**Comment by ValdasK** [comment]
Seems like something fishy happened, so just removed everything and it installed correctly this time.

```
DELETE FROM oc_migrations WHERE app = 'mail';
DROP TABLE oc_mail_accounts;
DROP TABLE oc_mail_aliases;
DROP TABLE oc_mail_coll_addresses;
DROP TABLE oc_mail_attachments;
DROP TABLE oc_mail_mailboxes;
```

**Comment by ChristophWurst** [comment]
Okay, thanks for the info.

Should this happen again with this or another app, be sure to save the full output and error logs!

**Comment by stellarpower** [comment]
I know I should open a separate issue, and if I have the itme, I will, but I can't right now and am just going to drop this here before I forget - I have had similar issues when I first fired up my nextcloud instance (using a docker-compose; fpm and postgres), and it sees now logging in via the app, and occasionally browsers. It seems to be caused by timeouts. I'm hosting the whole thing on a Raspberry Pi in my home for power-saving reasons, however it is not a fast machine. I believe I've set up all the proxies with sufficient timeout settings, perhaps not, however what I'm wondering is if it's caused by an internal timer issue
- upon first launch, I had to wait a good 20 minutes to be sure initialisation had been performed correctly upon setting up the root account; the page sat spinning away, and if I refreshed too soon, I got a similar error about tables already existing.
- With the mail app, I marked it to be downloaded and installed and perhaps  proceded too quickly before navigating to another page
- When I logged in on the android app, the flow hangs, waiting for me to enter my password, and then if I hit back, goes ahead to authorisation, where it also hangs; I had to use a QR code which worked fine.

In cases 1 and 2 I am seeing database issues and in all three cases my best guess is somewhere internally there's an assumption that my system won't take as long to respond as it does, and so an implicit timeout occurs when the database server for example is still halfway through applying the updates requested but the fpm instance assumes that the operation either must have completed successfully or terminated; we're not talking minutes for normal operations, it's perfectly usable under everyday browsing of files and notes, but at times I just have to be patient, which I myself am more than happy to do after such a long effort to get this up and running. 

TL;DR - IMO this specific error here *could* be a part of a larger, even if relatively simple, issue in other parts of nextcloud, and in my case I would believe that to be the case (same error message, and I presume #3136 will solve this once it's pulled in)
